EIPS/eip-7718.md

```diff
@@ -0,0 +1,418 @@
+
+- Request messages **MUST** be sent using a `TALKREQ` message.
+- Response messages **MUST** be sent using the corresponding `TALKRESP` message.
+
+All messages are encoded as an [SSZ Union](https://github.com/ethereum/consensus-specs/blob/04f5ec595d78c0e3e43794fb7644c18f2584770d/ssz/simple-serialize.md#union) type.
+
+```
+message = Union[ping, pong, find_nodes, nodes, find_content, content, offer, accept]
+serialized_message = serialize(message)
+```
+
+The `serialized_message` is the payload passed to the `request` field of the `TALKREQ` message or the `reponse` field of the `TALKRESP` message.
+
+The type values for the `Union` are the SSZ Containers specified per message type.
+
+The transmission of `content` data that is too large to fit a single packet is done over [uTP](../assets/eip-7718/bep_0029-rst_post.pdf).
+
+#### Aliases
+
+For convenience we alias:
+
+- `ByteList` to `List[uint8, max_length=2048]`
+
+#### Message Types
+
+##### Ping (0x00)
+
+Request message to check if a node is reachable, communicate basic information about our node, and request basic information about the recipient node.
+
+```
+selector     = 0x00
+ping         = Container(enr_seq: uint64, custom_payload: ByteList)
+```
+
+- `enr_seq`: The node's current sequence number of their ENR record.
+- `custom_payload`: Custom payload specified per the network.
+
+##### Pong (0x01)
+
+Response message to Ping(0x00)
+
+```
+selector     = 0x01
+pong         = Container(enr_seq: uint64, custom_payload: ByteList)
+```
+
+- `enr_seq`: The node's current sequence number of their ENR record.
+- `custom_payload`: Custom payload specified per the network.
+
+##### Find Nodes (0x02)
+
+Request message to get ENR records from the recipient's routing table at the given logarithmic distances. The distance of `0` indicates a request for the recipient's own ENR record.
+
+```
+selector     = 0x02
+find_nodes   = Container(distances: List[uint16, max_length=256])
+```
+
+- `distances`: a list of distances for which the node is requesting ENR records for.
+    - Each distance **MUST** be within the inclusive range `[0, 256]`
+    - Each distance in the list **MUST** be unique.
+
+##### Nodes (0x03)
+
+Response message to FindNodes(0x02).
+
+```
+selector     = 0x03
+nodes        = Container(total: uint8, enrs: List[ByteList, max_length=32])
+```
+
+- `total`: The total number of `Nodes` response messages being sent.
+- `enrs`: List of byte strings, each of which is an RLP encoded ENR record.
+    - Individual ENR records **MUST** correspond to one of the requested distances.
+    - It is invalid to return multiple ENR records for the same `node_id`.
+    - The ENR record of the requesting node **SHOULD** be filtered out of the list.
+
+> Note: If the number of ENR records cannot be encoded into a single message, then they should be sent back using multiple messages, with the `total` field representing the total number of messages that are being sent.
```

Comment:
I think we need to remove this note as we don't have a solution currently for this multiple messages while remaining discv5 compliant